src/protocol/templates/ping1d.cpp.in

```diff
@@ -0,0 +1,212 @@
+    if (!set_ping_interval(ping_interval_ms)) {
+        return false;
+    }
+
+    return true;
+}
+
+PingMessage* Ping1D::waitMessage(enum Ping1DNamespace::msg_ping1D_id id, uint16_t timeout_ms)
+{
+    uint32_t tstart = millis();
+    while (millis() < tstart + timeout_ms) {
+
+        PingMessage* pmsg = read();
+
+        if (!pmsg) {
+            continue;
+        }
+
+        handleMessage(pmsg);
+
+        if (pmsg->message_id() == Ping1DNamespace::Nack) {
+            ping_msg_ping1D_nack nack(*pmsg);
+
+            // TODO remove, this won't ever be true in the current design
+            if (nack.nacked_id() == id) {
+                return nullptr;
+            }
+        }
+
+        if (pmsg->message_id() == id) {
+            return pmsg;
+        }
+    }
+
+    return nullptr;
+}
+
+void Ping1D::handleMessage(PingMessage* pmsg)
+{
+    switch (pmsg->message_id()) {
+{% for msg in messages["ping1D"]["get"] %}
+        case Ping1DNamespace::{{msg|title}}:
+        {
+            ping_msg_ping1D_{{msg}} m(*pmsg);
+{% for field in messages["ping1D"]["get"][msg].payload %}
+{% if is_vector(field.type) %}
+            if (m.{{field.name}}_length() > _{{field.name}}_length) {
+                if (_{{field.name}}) {
+                    free(_{{field.name}});
+                }
+                _{{field.name}} = ({{get_type_string(field.vector.datatype)}}*)malloc(_{{field.name}}_length * sizeof({{get_type_string(field.vector.datatype)}}));
+            }
+
+            _{{field.name}}_length = m.{{field.name}}_length();
+
+            for (uint16_t i = 0; i < _{{field.name}}_length; i++) {
+                _{{field.name}}[i] = m.{{field.name}}()[i];
+            }
+{% else %}
+            _{{field.name}} = m.{{field.name}}();
+{% endif %}
+{% endfor %}
+        }
+            break;
+
+{% endfor %}
+
+        default:
+            break;
+    }
+}
+
+// ex ping_msg_ping1D_voltage_5 msg(*pd.request(Ping1DNamespace::Voltage_5));
+PingMessage* Ping1D::request(enum Ping1DNamespace::msg_ping1D_id id, uint16_t timeout_ms)
+{
+    ping_msg_ping1D_empty msg;
+    msg.set_id(id);
+    msg.updateChecksum();
+    write(msg.msgData, msg.msgDataLength());
+    return waitMessage(id, timeout_ms);
+}
+
+// ex auto msg = pd.request<ping_msg_ping1D_voltage_5>();
+template <typename T>
+T* Ping1D::request()
+{
+    T resp; // todo there should be some other (static) way to get the message id?
+    static ping_msg_ping1D_empty req;
+    req.set_id(resp.message_id());
+    req.updateChecksum();
+    write(req.msgData, req.msgDataLength());
+    return (T*)waitMessage(resp.message_id());
+}
+
+
+{% for msg in messages["ping1D"]["get"] %}
+bool Ping1D::get_{{msg}}(
+{%- for field in messages["ping1D"]["get"][msg].payload %}
+{% if is_vector(field.type) %}
+{{ "                     " if not loop.first }}{{get_type_string(field.vector.sizetype)}}* {{field.name}}_length, {{get_type_string(field.vector.datatype)}}** {{field.name}}{{ ",\n" if not loop.last }}
```

Comment:
Why not reference ?
Using pointers will make it not clear for the user, and is less dumb proof.
Right now it's necessary to move all variables to pointer:
```
  if (ping.get_general_info(
            &version_major,
            &version_minor,
            &mvolts,
            &msec_per_ping,
            &gain_index,
            &is_auto))
  {
```
Moving to reference will probably make it simpler:
```
  if (ping.get_general_info(
            version_major,
            version_minor,
            mvolts,
            msec_per_ping,
            gain_index,
            is_auto))
  {
```

Reply:
Besides that, the compiler will not fail if the user sends int as arguments, the compiler will thing that the values are memory address.
```cpp
int foo1(int* foo) { return 0; };
int foo2(int& foo) { return 0; };

void loop() {
  // No error
  foo1(1);
  // Will fail to compile
  foo2(1);
}
```

Reply:
We discussed this. Reference needs lvalue and initialization. We will need to add static dummy variable or other tricks to get around the language impositions with references. This is the recommended way to return multiple values with c++.

Reply:
> Using pointers will make it not clear for the user

We have clear api documentation and examples.

If a user is trying to 

`get_distance(1);`

That's silly enough that it doesn't concern me. Maybe there is something else we can put here for better type safety.

The REAL solution here would to be return a payload struct. But we do not have these except in the struct branch of protocol...

Reply:
I forgot that we have already discussed that.